references/depth/stereo/utils/losses.py

```diff
@@ -0,0 +1,477 @@
+        self.rescale_factor = rescale_factor
+        self.rescale_mode = rescale_mode
+
+    def forward(self, flow_preds: List[Tensor], weights: Optional[Tensor] = None) -> Tensor:
+        """
+        Args:
+            flow_preds: list of tensors of shape (batch_size, C, H, W)
+
+        Returns:
+            sequence consistency loss of shape (batch_size,)
+        """
+        return _flow_sequence_consistency_loss_fn(
+            flow_preds,
+            gamma=self.gamma,
+            resize_factor=self.resize_factor,
+            rescale_factor=self.rescale_factor,
+            rescale_mode=self.rescale_mode,
+            weights=weights,
+        )
+
+
+def _psnr_loss_fn(source: torch.Tensor, target: torch.Tensor, max_val: float) -> torch.Tensor:
+    torch._assert(
+        source.shape == target.shape,
+        "psnr_loss: source and target must have the same shape, but got {} and {}".format(source.shape, target.shape),
+    )
+
+    # ref https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio
+    return 10 * torch.log10(max_val**2 / ((source - target).pow(2).mean(axis=(-3, -2, -1))))
+
+
+class PSNRLoss(nn.Module):
+    def __init__(self, max_val: float = 256) -> None:
+        """
+        Args:
+            max_val: maximum value of the input tensor. This refers to the maximum domain value of the input tensor.
+
+        """
+        super().__init__()
+        self.max_val = max_val
+
+    def forward(self, source: Tensor, target: Tensor) -> Tensor:
+        """
+        Args:
+            source: tensor of shape (D1, D2, ..., DN, C, H, W)
+            target: tensor of shape (D1, D2, ..., DN, C, H, W)
+
+        Returns:
+            psnr loss of shape (D1, D2, ..., DN)
+        """
+
+        # multiply by -1 as we want to maximize the psnr
+        return -1 * _psnr_loss_fn(source, target, self.max_val)
+
+
+class FlowPhotoMetricLoss(nn.Module):
+    def __init__(
+        self,
+        ssim_weight: float = 0.85,
+        ssim_window_size: int = 11,
+        ssim_max_val: float = 1.0,
+        ssim_sigma: float = 1.5,
+        ssim_eps: float = 1e-12,
+        ssim_use_padding: bool = True,
+        max_displacement_ratio: float = 0.15,
+    ) -> None:
+        super().__init__()
+
+        self._ssim_loss = SSIM(
+            kernel_size=ssim_window_size,
+            max_val=ssim_max_val,
+            sigma=ssim_sigma,
+            eps=ssim_eps,
+            use_padding=ssim_use_padding,
+        )
+
+        self._L1_weight = 1 - ssim_weight
+        self._SSIM_weight = ssim_weight
+        self._max_displacement_ratio = max_displacement_ratio
+
+    def forward(
+        self,
+        source: Tensor,
+        reference: Tensor,
+        flow_pred: Tensor,
+        valid_mask: Optional[Tensor] = None,
+    ):
+        """
+        Args:
+            source: tensor of shape (B, C, H, W)
+            reference: tensor of shape (B, C, H, W)
+            flow_pred: tensor of shape (B, 2, H, W)
```

Comment:
Question, Is it possible to make this loss more generic so it can be used for 1-channel disparity, i.e flow_pred shape (B, 1, H, W) ?

Not sure if this is correct, but here is my idea for this:
```
Current:
  max_H = int(H * self._max_displacement_ratio)
  max_W = int(W * self._max_displacement_ratio)
New:
  if flow_pred.shape[1] == 1:
    max_displacements = [int(W * self._max_displacement_ratio)]
  elif flow_pred.shape[2] == 2:
    max_displacements = [int(H * self._max_displacement_ratio), int(W * self._max_displacement_ratio)]
  else:
    raise ValueError(f"flow_pred.shape[1] should be either 1 or 2, but got {flow_pred.shape[1]}")
```

```
Current:
 max_flow_mask = torch.logical_and(
     flow_pred[:, 0, :, :] < max_W,
     flow_pred[:, 1, :, :] < max_H,
  )

New:
  max_flow_mask = torch.logical_and(*[
      flow_pred[:, i, :, :] < max_displacements[i] 
      for i in range flow_pred.shape[1] 
  ])
```